### ISSUE TITLE
Questions on the difference between the ARM and muRISCV implementations in nn_mat_mult_kernel_s8_s16

### ISSUE THREAD

**Issue by tomhepworth** [tum-ei-eda/muriscv-nn#64]
Hi, I have a few things I am trying to understand with the nn_mat_mult_kernel_s8_s16 function. 

Firstly, I see the arm code accumulates four values in [its inner loop](https://github.com/ARM-software/CMSIS-NN/blob/main/Source/ConvolutionFunctions/arm_nn_mat_mult_kernel_s8_s16.c#L116). But in the muriscv repo the loop seems to [use eight accumulators](https://github.com/tum-ei-eda/muriscv-nn/blob/master/Source/ConvolutionFunctions/muriscv_nn_mat_mult_kernel_s8_s16.c#L193). Is there a reason why this was changed? I am also confused because the extra channels do not seem to be accumulated into in the loop computing the `remaining` rows. 

Perhaps you could also provide some clarification on the `//new` comments and what they are referring to? 

Additionally, I observe that all of the values used in my model which runs this function are int8. However the code uses the q15 type. Why is this? Is it for sign extension or for widening before the mul? 

PS. I am trying to understand this algorithm generally, but there is not a great deal of documentation available. I understand that this function is used in im2col convolution, and it somehow caches the im2col image and processes it in smaller chunks. If you have any materials to help me understand what is going on I would be very grateful if you could provide them. Thanks :)

**Comment by PhilippvK** [comment]
@tomhepworth Thank you for that question. My colleague who came up with this change won't be available for another week, so I can not give you a definitive answer.

If I remember correctly, we have been doing comparisons between our manually vectorized kernels compared to the scalar kernels vectorized by LLVM and found some cases where muRISCV-NN Scalar actually yielded better performance than muRISCV-NN Vector. After some investigations, we saw that LLVM did unroll the kernel more aggressively and we ported that to your manually vectorized kernels. @ParkerJones567 please correct me if I am wrong in this specific case.

As RISC-V has a larger register file than ARM (31 vs less than 16 available registers), using a few more registers here should not be an Issue for muRISCV-NN, while it would probably be problematic for CMSIS-NN.

**Comment by tomhepworth** [comment]
Okay, so I should be able to edit the function on a branch so that it only uses four again and still match the behaviour of CMSIS?

**Comment by PhilippvK** [comment]
I guess that would work.

**Comment by ParkerJones567** [comment]
Hello, I can try to answer your questions for you.

This function is only used by scalar and P extension implementations (the outer level  "#if !defined(USE_VEXT)" only includes the code for these configurations, otherwise it returns NULL).

The reason for 8 accumulators here is that the output, after requantization, is an 8-bit value.  Using the RISCV Packed extension, four int 8 values can be packed in a single 32 bit register, allowing them all to be read/written at the same time.  The original CMSIS kernel uses 4 accumulators, 2 unrolled rows with 2 accumulators per row.  By unrolling to 4 rows, performance (number of cycles in an ISS) was improved by reducing the number of required reads/writes.  It was also observed to help with the scalar implementation so it was added there as well.

This change was not included in the remaining rows loop as our current benchmarks never enter this part of the loop (all benchmarks that utilize this functions have a number of rows that are a multiple of 4).  It is likely to improve performance there as well, but as it can't be tested/verified it has not been changed.

Regarding the use of the q15 type, this function is used within the muriscv_nn_convolve_s8 kernel and is written to accept 'input_b' as a *q15 (thus the _s8_s16 in the name of the function).  Changing this would require further changes to muriscv_nn_convolve_s8 to handle this case.  The im2col transform also takes place in this kernel, along with other special cases, which may help you in trying to figure out how they all work.

Regarding the '//new' comments, these are just leftover from the last time these kernels were modified, sorry about that.

Hopefully I answered everything, feel free to ask further questions!
